crates/bevy_ecs/src/system/function_system.rs

```diff
@@ -50,6 +50,21 @@ pub struct SystemMeta {
     pub(crate) commands_span: Span,
 }
 
+/// Trait for retrieving system metas.
+pub trait SystemMetaProvider {
+    /// Returns immutable references of system metas as a vector.
+    fn system_metas(&self) -> Vec<&SystemMeta>;
+
+    /// Returns mutable references of system metas as a vector.
+    fn system_metas_mut(&mut self) -> Vec<&mut SystemMeta>;
+
+    /// Helper method for `system_metas`.
+    fn extend_with_system_metas<'a>(&'a self, vec: &mut Vec<&'a SystemMeta>);
```

Comment:
it might be better for these to take an `impl Extend` instead of a Vec. https://doc.rust-lang.org/core/iter/trait.Extend.html

Reply:
It's neat, but I don't think it's appropriate to implement it here

Reply:
I was saying to change the function signature to `fn extend_with_system_metas<'a>(&'a self, extendable: impl Extend<&'a SystemMeta>)`, which would allow the methods to take a bunch of the other std collections and not be limited to just a `Vec`.

Reply:
Another option is to take `impl FnMut(&'a SystemMeta)`.  That way, users that are just going to iterate over the result won't even have to allocate a `Vec`.

Reply:
Considering we want to be able to map items and do other sorts of behaviour for the metas in the follow ups, this could be more flexible.

Reply:
Yup

Reply:
Just to clarify, my point was that `FnOnce` is more general than `Extend` here.  You'd still offer the version that returns a `Vec`!  

```rust
fn for_each_system_meta<'a>(&'a self, f: impl FnOnce(&'a SystemMeta));

fn system_metas(&self) -> Vec<&SystemMeta> {
    let mut result = Vec::new();
    self.for_each_system_meta(|meta| result.push(meta));
    result
}
```
If someone wants to populate another type of collection, they can pass a closure that calls the relevant `push` method.  But if they were just going to do `for meta in system.system_metas() { ... }` then they can avoid an allocation by calling `system.for_each_system_meta(|meta| ...)`.

Reply:
Can't be `FnOnce` since we use it many times, besides that makes sense, I voice no preference then

Reply:
Idea seems reasonable.